addons/phone_validation/models/phone_blacklist.py

```diff
@@ -40,15 +47,33 @@ def create(self, values):
             done.add(sanitized)
             to_create.append(dict(value, number=sanitized))
 
-        """ To avoid crash during import due to unique email, return the existing records if any """
-        sql = '''SELECT number, id FROM phone_blacklist WHERE number = ANY(%s)'''
-        numbers = [v['number'] for v in to_create]
-        self._cr.execute(sql, (numbers,))
-        bl_entries = dict(self._cr.fetchall())
-        to_create = [v for v in to_create if v['number'] not in bl_entries]
-
-        results = super(PhoneBlackList, self).create(to_create)
-        return self.env['phone.blacklist'].browse(bl_entries.values()) | results
+        # Check existing records for the numbers requested to create
+        sql = '''SELECT number, id, active FROM phone_blacklist WHERE number = ANY(%s)'''
+        self._cr.execute(sql, ([tc['number'] for tc in to_create],))
+        class ResultLine(NamedTuple):
+            number: str
+            record_id: int
+            active: bool
+
+        already_active_ids = set()
+        to_activate_ids = set()
+        existing_numbers = set()
+        for rl in starmap(ResultLine, self._cr.fetchall()):
+            existing_numbers.add(rl.number)
+            if rl.active:  # record already exists and is active -> no need to create it again
+                already_active_ids.add(rl.record_id)
+            else:
+                to_activate_ids.add(rl.record_id)
+
+        # Create new records, while skipping existing_numbers
+        to_create = [tcr for tcr in to_create if tcr['number'] not in existing_numbers]
+        created = super().create(to_create)
+
+        # Blacklist unblacklisted (i.e activate nonactive)
+        activated = self.env['phone.blacklist'].browse(to_activate_ids)
+        activated.write({'active': True})
```

Comment:
As much as I appreciate itertools usage, this seems awfully complicated just to sort some records ^^
Typically I'd say if you end up playing with "sets of ids", there's a good chance you want to be using record sets.

Can we maybe use search_fetch with `sudo` as `search_fetch([('active', 'in', [True, False]), ('number', 'in', numbers)], ['active', 'number'])` and then just filter normally?
like just `already_active = black_list_entries.filtered('active')`

I assume the reason it's not done here is that search_fetch is relatively recent.

Reply:
I don't see `search_fetch` in **16.0**  I though I will assemble it myself, but I also don't see _fetch_query
I'm transitioning to existing ORM methods (removing out sql )

Reply:
Ah indeed, goal was to ensure we don't add queries but indeed just a regular search is fine.